pkg/cvo/sync_worker.go

```diff
@@ -431,47 +441,42 @@ func (w *SyncWorker) apply(ctx context.Context, payloadUpdate *payload.Update, w
 			Backoff:  w.backoff,
 		})
 	}
+	graph := payload.NewTaskGraph(tasks)
+	graph.Split(payload.SplitOnJobs)
+	graph.Parallelize(payload.ByNumberAndComponent)
 
-	for i := 0; i < len(tasks); i++ {
-		task := tasks[i]
-		setAppliedAndPending(version, total, done)
-		fraction := float32(i) / float32(len(tasks))
-
-		reporter.Report(SyncWorkerStatus{Generation: work.Generation, Fraction: fraction, Step: "ApplyResources", Reconciling: work.Reconciling, VersionHash: payloadUpdate.ManifestHash, Actual: update})
-
-		glog.V(4).Infof("Running sync for %s", task)
-		glog.V(5).Infof("Manifest: %s", string(task.Manifest.Raw))
-
-		if contextIsCancelled(ctx) {
-			err := fmt.Errorf("update was cancelled at %d/%d", i, len(tasks))
-			reporter.Report(SyncWorkerStatus{Generation: work.Generation, Failure: err, Fraction: fraction, Step: "ApplyResources", Reconciling: work.Reconciling, VersionHash: payloadUpdate.ManifestHash, Actual: update})
-			return err
-		}
+	// update each object
+	err := payload.RunGraph(ctx, graph, maxWorkers, func(ctx context.Context, tasks []*payload.Task) error {
+		for _, task := range tasks {
+			if contextIsCancelled(ctx) {
+				return cr.CancelError()
+			}
+			cr.Update()
 
-		ov, ok := getOverrideForManifest(work.Overrides, task.Manifest)
-		if ok && ov.Unmanaged {
-			glog.V(4).Infof("Skipping %s as unmanaged", task)
-			continue
-		}
+			glog.V(4).Infof("Running sync for %s", task)
+			glog.V(5).Infof("Manifest: %s", string(task.Manifest.Raw))
 
-		if err := task.Run(version, w.builder); err != nil {
-			reporter.Report(SyncWorkerStatus{Generation: work.Generation, Failure: err, Fraction: fraction, Step: "ApplyResources", Reconciling: work.Reconciling, VersionHash: payloadUpdate.ManifestHash, Actual: update})
-			cause := errors.Cause(err)
-			if task.Requeued == 0 && shouldRequeueOnErr(cause, task.Manifest) {
-				task.Requeued++
-				tasks = append(tasks, task)
+			ov, ok := getOverrideForManifest(work.Overrides, task.Manifest)
+			if ok && ov.Unmanaged {
+				glog.V(4).Infof("Skipping %s as unmanaged", task)
 				continue
 			}
-			return err
+
+			if err := task.Run(version, w.builder); err != nil {
+				return err
+			}
+			cr.Inc()
+			glog.V(4).Infof("Done syncing for %s", task)
 		}
-		done++
-		glog.V(4).Infof("Done syncing for %s", task)
+		return nil
+	})
+	if err != nil {
+		cr.Error(err)
+		return err
 	}
 
-	setAppliedAndPending(version, total, done)
 	work.Completed++
```

Comment:
why are we tracking completed in both `work` and `consistentReporter` ?

Reply:
One is updating the status we send back to the main CVO loop (via reporter), one is used internally by the sync_worker to report for the next generation.  I moved this line up to where it belongs, in Start()